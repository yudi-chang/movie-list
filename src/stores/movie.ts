import { defineStore } from 'pinia';
import type { Movie } from '@/types';

type MovieState = {
  favorites: Movie[];
}

export const useMovieStore = defineStore('movie', {
  state: (): MovieState => ({
    favorites: [],
  }),
  actions: {
    addFavorites(movie: Movie) {
      this.favorites.push(movie)
    },
    removeFromFavorites(moviePayload: Movie) {
      const index = this.favorites.findIndex(movie => movie.imdbID === moviePayload.imdbID);
      if (index !== -1) {
        this.favorites.splice(index, 1);
      }
    }
  },
});
