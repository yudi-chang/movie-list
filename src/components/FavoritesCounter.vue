<template>
  <div class="counter-wrapper text-xs">
    {{ movieStore.favoritesCount }}
  </div>
</template>

<script setup lang="ts">
import { useMovieStore } from '@/stores/movie';

// store
const movieStore = useMovieStore();
</script>

<style scoped lang="scss">
.counter-wrapper {
  border-radius: 5px;
  padding: 0 5px;
  position: absolute;
  right: -10px;
  top: 0;
  background-color: #e74c3c;
}


@media (max-width: $breakpoint-sm) {
  .counter-wrapper {
    right: -15px;
    top: -20px;
  }
}
</style>