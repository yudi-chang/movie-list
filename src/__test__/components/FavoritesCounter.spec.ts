import { describe, it, expect, beforeEach } from 'vitest'
import { mount } from '@vue/test-utils'
import { createPinia, setActivePinia } from 'pinia';
import { useMovieStore } from '@/stores/movie';
import FavoritesCounter from '@/components/FavoritesCounter.vue'
import movieResponse from '@/__test__/mocks/get-movies'

describe('FavoritesCounter', () => {
  const defaultParams = {};

  const createWrapper = (params = defaultParams) => {
    return mount(FavoritesCounter, params)
  }

  beforeEach(() => {
    setActivePinia(createPinia());
  });

  it('should render counter correctly', async () => {
    const movieStore = useMovieStore();
    const wrapper = createWrapper();

    movieStore.addFavorites(movieResponse.data[0])
    movieStore.addFavorites(movieResponse.data[1])
    movieStore.addFavorites(movieResponse.data[2])
    await wrapper.vm.$nextTick()

    expect(wrapper.find('[data-test="counter-wrapper"]')).toMatchSnapshot();
  })
})
