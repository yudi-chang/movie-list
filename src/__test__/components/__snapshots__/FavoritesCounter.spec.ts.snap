// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`FavoritesCounter > should render counter correctly 1`] = `
DOMWrapper {
  "isDisabled": [Function],
  "wrapperElement": <div
    class="counter-wrapper text-xs"
    data-test="counter-wrapper"
    data-v-f9cbd204=""
  >
    3
  </div>,
}
`;
